bip-0340.mediawiki

```diff
@@ -164,7 +164,7 @@ The algorithm ''Sign(sk, m)'' is defined as:
 
 ==== Alternative Signing ====
 
-It should be noted that various alternative signing algorithms can be used to produce equally valid signatures. The algorithm in the previous section is deterministic, i.e., it will always produce the same signature for a given message and secret key. This method does not need a random number generator (RNG) at signing time and is thus trivially robust against failures of RNGs. Alternatively the 32-byte ''rand'' value may be generated in other ways, producing a different but still valid signature (in other words, this is not a ''unique'' signature scheme). '''No matter which method is used to generate the ''rand'' value, the value must be a fresh uniformly random 32-byte string which is not even partially predictable for the attacker.'''
+It should be noted that various alternative signing algorithms can be used to produce equally valid signatures. The algorithm in the previous section is deterministic, i.e., it will always produce the same signature for a given message and secret key. This method does not need a random number generator (RNG) at signing time and is thus trivially robust against failures of RNGs. Alternatively the 32-byte ''rand'' value may be generated in other ways, producing a different but still valid signature (in other words, this is not a ''unique'' signature scheme). '''No matter which method is used to generate the ''rand'' value, the value must be a fresh uniformly random 32-byte string which is not even partially predictable for the attacker.''' Freshness implies that reusing a secret key in different signature schemes is discouraged. For example, if the ''rand'' value was computed as per RFC6979 and the same secret key is used in deterministic ECDSA with RFC6979, the signatures can leak the secret key through nonce reuse.
```

Comment:
Useful advice.

Reply:
Great advice.  I'm not a fan of the phrasing "is discouraged".  It feels like an RFC6919 "MUST (BUT WE KNOW YOU WON'T)". 

"Nonce freshness with a derandomize nonce function implies that the same inputs must not be presented in another context. This can be most reliably accomplished by not reusing the same private key across different signing schemes. For example,"

Reply:
I think you mean that the *nonce* must be not even partially predictable for the attacker. No harm is done if *rand* is partially predictable (or entirely predictable), if the prescribed algorithm is used to generate the nonce using it.

Reply:
nonce is maybe ambigious.  for example, use RFC6979 with two different curves.  "R" will be different and your secret key will be leaked.

Reply:
Yeah, and someone might think a linearly related nonce to another nonce is unpredictable if the other one was. I don't know how to best describe this.

I think the property we should aim to encourage is that nonce generating function(s), ignoring any randomness fed to them, should be secure. That means either no keys reused across schemes, or different tag/nonce function. Adding randomness to it is to prevent fragility against implementation bugs, or against fault attacks, but shouldn't be treated as a first layer of security.

Reply:
@sipa It seems like you're confusing the auxiliary input to the nonce function (i.e. synthetic nonces, see also here https://github.com/bitcoin/bips/pull/886) with the `rand` variable that is the output of the deterministic nonce function (as defined a few lines above). This PR only affects the latter.

Reply:
Oops, I was reviewing on mobile and missing the context it seems.